Move to v2 of the 3scale go client

philipgough opened the pull request:
This change is the first of several commits which will introduce
breaking changes for the old client API's.

Here we move across the types into a new package whose import path
begins to make more sense than what existed
previously ('3scale-go-client/client').

The reason fro introducing this breaking change is usability.
We have taken our learnings from the using this library the dogfooding and will
hopefully drastically improve the user experience for others going
forward.

Further to that, we have, while working in other projects that rely on
this code, discovered use cases that were not met by the previously
defined exported functions.

Finally, we introduce an interface here for 3scale backend 'Backend'
which will solidfy the API and avoid us having to write, repeatedly,
fake network responses in multiple projects.

philipgough committed:
Move to v2 of the 3scale go client

This change is the first of several commits which will introduce
breaking changes for the old client API's.

Here we move across the types into a new package whose import path
begins to make more sense than what existed
previously ('3scale-go-client/client').

The reason fro introducing this breaking change is usability.
We have taken our learnings from the using this library the dogfooding and will
hopefully drastically improve the user experience for others going
forward.

Further to that, we have, while working in other projects that rely on
this code, discovered use cases that were not met by the previously
defined exported functions.

Finally, we introduce an interface here for 3scale backend 'Backend'
which will solidfy the API and avoid us having to write, repeatedly,
fake network responses in multiple projects.

philipgough committed:
Expose useful helper functions for package types

This change exposes various functions that are useful to the caller of
Backend API.

Adds internal functions for validation, convrsions and formatting.

philipgough committed:
test: Unit test for helper/constructor functions

philipgough committed:
Adds Option functions to wrap Request in extended functionalities

philipgough committed:
Add Authorize API

This change introduces the restructured Authorize API and adds the
signature to the Backend interface.

philipgough committed:
test: Unit test for Authorize endpoint

philipgough committed:
Add AuthRep API

philipgough committed:
test: Unit test for AuthRep

This is significantly minimised in comparison to the Authorize test
because the code follows the same path.

philipgough committed:
Remove client side validation

After some discussion, we decided to remove the client side validation
from incoming user input.

The reason for these changes are to allow for server side changes
without the need for tying the client version to specific and strict
policies.

While the changes here had prevented input from returning responses that
we now known would result in 403 errors, the tradeoff for removal is
felt warranted.

This also allows other tokens outside those defined via the iota
to be provided by the caller.

philipgough committed:
test: Update unit test to reflect removal of validation

philipgough committed:
Adds Report API

Report API now supports reporting batches transactions in a single call.

Documents the ReportResponse return type as well as adding an struct
which supports decoding reporting errors.

Extends the Backend interface to support Reporting.

A helper function has been added to convert incoming Report transactions
to correctly formatted query URL. This is necessary because the URL
formatting expected by 3scale is different than that of Auth and AuthRep because of
the batching support.

philipgough committed:
test: Unit test for reporting

philipgough committed:
Deprecate old client APIs

This change totally removes the old APIs, their associated types and
tests.

philipgough committed:
Update Makefile to test new package

philipgough committed:
Rename Request to Transaction

To avoid any confusion between HTTP requests and '3scale Requests',
this change renames the Request type to Trnasaction. This is also more
inkeeping with existing 3scale documentation.

philipgough committed:
DeepCopy functions for Metrics and Hierarchy

This change adds utility functions for cloning the map structures.

philipgough committed:
test: Tests for DeepCopy functions

philipgough committed:
mod: Create module

unleashed commented on threescale/api.go:
```diff
@@ -94,6 +113,34 @@ func (c *Client) doAuthorizeReq(req *http.Request, extensions Extensions) (*Auth
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusAccepted {
```
Best to take any 2XX value in these cases, not just 202.

PhilipGough replied:
yep had thought about that. Will fix

unleashed commented on threescale/types.go:
```diff
@@ -154,6 +144,16 @@ type UsageReport struct {
+	Params     Params
+	Timestamp  string
+	context    context.Context
+	extensions Extensions
```
These are currently just per request and it's unlikely they will ever be per transaction because it would end up wasting lots of space (I think you change this in another commit, though).

PhilipGough replied:
I don't fix it actually but had thought about it. So I wanted a way to surface some obscure functionality to the end user. I also wanted to avoid bloating the API for simple use cases, but maybe we need to accept `*Options` in the interface methods, wdyt?

unleashed replied:
Yes, some mechanism to specify these per request which you don't need to bother (not too much anyway) if you are not using it.

unleashed commented on threescale/common.go:
```diff
@@ -111,6 +123,15 @@ func (m Metrics) Delete(name string) {
 }
 
+// DeepCopy returns a clone of the original Metrics
+func (m Metrics) DeepCopy() Metrics {
```
Do we have any place where we can use these deep copies ourselves?

PhilipGough replied:
Yes, in the adapter we are copying these objects. I put them here for that reason.

unleashed replied:
:+1:

unleashed reviewed (commented):
Added some comments on previous commits that were changed on dropping validation, but good to check them out.

PhilipGough commented on threescale/common.go:
```diff
@@ -0,0 +1,266 @@
+func (ca ClientAuth) validate() error {
+	if ca.Type != 0 && ca.Type != 1 {
+		return fmt.Errorf("auth is invalid, unknown token type provided")
+	}
```
Yes this was changed later based on our discussion over slack on keeping it generic. I was mimicking enums here but moved away from that too static strings

PhilipGough commented on threescale/common.go:
```diff
@@ -0,0 +1,266 @@
+func (m Metrics) convert() map[string]string {
+	formatted := make(map[string]string, len(m))
+	for k, v := range m {
+		if v > 0 {
```
If they choose to use the utility methods yes. But not if it was created raw. I can remove the check here if you want though, that might make sense. The utility methods will provide some validation but can be bypassed by creating metrics directly in that case.

Does that sound reasonable?

unleashed replied:
Yes, I think so.

PhilipGough replied:
Done. Keep validation in the `Add`, `Set` methods but provide a workaround for someone who doesn't care

PhilipGough commented on threescale/common.go:
```diff
@@ -0,0 +1,266 @@
+		}
+	}
+	return values
+}
```
Yep understood. If you are happy enough with this I would like to leave it. The trade off is indeed performance but we know the max size of this object upfront is not very large and so iterating over it is not all that costly IMO.

I'm happy to revisit in another PR later if we want some static conversion. Does that sound ok?

unleashed replied:
Sounds good!

PhilipGough commented on threescale/common.go:
```diff
@@ -0,0 +1,266 @@
+	if err == nil {
+		scheme := backendURL.Scheme
+		if scheme != "" && scheme != "http" && scheme != "https" {
+			err = fmt.Errorf("unsupported schema %s passed to backend", scheme)
```
the naming here matches the spelling on the underlying URL field that we match against so I think replacing the spelling is confusing. https://golang.org/pkg/net/url/#URL

unleashed replied:
Wdym? I don't see schema anywhere in that doc?

PhilipGough replied:
ah sorry, i see the typo now. I thought you were suggesting I rename the `scheme` var to `schema`. Apologies

PhilipGough commented on threescale/types.go:
```diff
@@ -59,7 +59,7 @@ type AuthorizeResponse struct {
 // AuthType maps to a known client authentication pattern
 // Currently known and supported are 0=ServiceToken 1=ProviderKey
-type AuthType int
+type AuthType string
```
The problem is the conversion to query params. How would we offer that? So a user passes a given iota and that needs to translate somehow to a string. When we are providing a fixed value up front that work but not in the case where we can't map the iota to a value right?

I think moving to string is fine and simplifies it but open to suggestion for sure.

unleashed replied:
It is fine, I was thinking more in line with the client knowing it is not using any of the known authentication types, but perhaps no code depends on it (ie. some optional validation could ensure the known types validate against known formats, but since it is now out it is probably not of use).

PhilipGough replied:
Yes, well they will know because they will need to cast their string as an AuthType to use anything outside `ServiceToken` and `ProviderKey`

PhilipGough commented on threescale/common.go:
```diff
@@ -0,0 +1,266 @@
+		baseURL:     defaultBackendUrl,
+		httpClient:  defaultHttpClient(),
+	}
+}
```
yes

PhilipGough replied:
+1

PhilipGough commented on threescale/common.go:
```diff
@@ -0,0 +1,266 @@
+
+// Add takes a provided key and value and adds them to the Metric 'm'
+// If the metric already existed in 'm', then the value will be added (if positive) or subtracted (if negative) from the existing value.
+// If a subtraction leads to a negative value Add returns an error  and the change will be discarded.
```
Resolved below by removing the validation from the conversion. So a user creating metrics via `m := Metrics{"403-risk": -5}` are free to do so

unleashed commented on threescale/api.go:
```diff
@@ -123,7 +123,8 @@ func (c *Client) doReportReq(req *http.Request, values url.Values) (*ReportRespo
 
-	if resp.StatusCode != http.StatusAccepted {
+	// ensure response is in 2xx range
+	if !(resp.StatusCode >= http.StatusOK && resp.StatusCode < http.StatusMultipleChoices) {
```
I'd expect there would be some helper to do this rather than effectively knowing the relation between these constants and their relative numbers and semantics. This is really not much better if at all than encoding the numbers directly. It is even probably worse from a semantics POV.

philipgough committed:
Restructure API to remove Options from Transaction

This chnage was made becasue it makes more sense to have the Options as
a standalone structure and not pass them as a field within each
Transaction.

This will now allow the user to provide n amount of available Options to
extend the functionality of the API, while keeping the general and
default use cases simple.

Because we no longer need the functional options on the Transaction, the
API has changed from accepting pointers to Transactions to Transaction
structs. The is more robust since it will not cause the library to panic
in cases where nil is passed by the user.

philipgough committed:
test: Update tests to reflect changes to Options in API

unleashed commented on threescale/api_test.go:
```diff
@@ -38,7 +39,7 @@ func TestClient_Authorize(t *testing.T) {
 			name:         "Test expect failure bad url passed",
 			auth:         ClientAuth{Type: ProviderKey, Value: "any"},
-			transaction:  &Transaction{Params: Params{AppID: "any"}},
+			transaction:  Transaction{Params: Params{AppID: "any"}},
```
Any test with a completely empty transaction meaning no transactions for the authorization call in order to retrieve the current limits state?

PhilipGough replied:
Yes covered here https://github.com/3scale/3scale-go-client/pull/23/files#diff-1f667ddd6c5e148c86dc54aa68c5147aR200-R229.

I'll verify that we dont generate usage data in said test.

philipgough committed:
test: Ensure no usage is generated for empty metric transaction

philipgough committed:
Use 3scale 'code' as Reason in Auth/AuthRep response

This change was made because often times, the 'Reason' returned from
3scale is an empty string. This provides no context to the end user as
to the cause of the error.

The 'Code' field is always populated and returns some brief information
on why the request failed.

Futhermore, the 'Reason', if it is populated often leaks out secrets and
this reduces the risk of such secrets getting entered into logging
systems.

unleashed commented:
@philipgough is this you resolving these queued jobs after ~7y?
